Dynamically update classloader after adding a library with the add-library command

OndroMih opened the pull request:
Dynamically updates the common classloader if a common library is added. It means that restart is not needed to add the library to the classpath.

OndroMih committed:
Update classloader in the add-library command

Dynamically updates the common classloader if a common library is added.

dmatej commented on nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/CommonClassLoaderServiceImpl.java:
```diff
@@ -155,10 +153,32 @@ public ClassLoader getCommonClassLoader() {
+     * @throws UnsupportedOperationException If adding not supported by the classloader
+     */
+    public void addToClassPath(URL url) {
+        if (commonClassLoader != null) {
```
Just invert the condition, `== null` should be first.

dmatej commented:
Is it possible also to remove the library? I see there is the `asadmin remove-library` command too, however I have no idea how it works if the application is already running and using the library.

pzygielo commented on nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/CommonClassLoaderServiceImpl.java:
```diff
@@ -155,10 +153,32 @@ public ClassLoader getCommonClassLoader() {
         return commonClassPath;
     }
 
+    private static String urlsToClassPath(Stream<URL> urls) {
```
What a great candidate to have unit test for!

pzygielo commented on nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/CommonClassLoaderServiceImpl.java:
```diff
@@ -155,10 +153,32 @@ public ClassLoader getCommonClassLoader() {
+        return urls
+                .map(URL::getFile)
+                .filter(Predicate.not(String::isBlank))
+                .map(file -> new File(file).getAbsolutePath())
```
Could the lambda be replaced with additional `.map`?

OndroMih replied:
You mean have 2 map calls instead of just one? It could but I thought it’s an unnecessary overhead to call map twice.

pzygielo replied:
> You mean have 2 map calls instead of just one? It could but I thought it’s an unnecessary overhead to call map twice.

Yes, yes.

2 map calls instead of one, and no lambda instead of one.

avpinchuk commented:
> Is it possible also to remove the library? I see there is the `asadmin remove-library` command too, however I have no idea how it works if the application is already running and using the library.

Perhaps, on Linux (and Unix generally) library file will be removed, but library will be accessible until server stop/restart. On Windows `remove-library` command should be failed due to file locking.

OndroMih commented:
The remove-library command just deletes the JAR on the file system. It could remove the url in the classloader but it’s not necessary. The CL will just ignore the url if the file doesn’t exist. So no need to remove the url from the CL. Classes loaded from the removed JAR will keep being loaded even after the JAR is removed.

avpinchuk commented:
> The CL will just ignore the url if the file doesn’t exist

If JAR was already opened by CL, CL will not ignore this JAR. Application will load classes and resources from this JAR.

dmatej commented:
We already have this command, so it should probably do more, ie. refuse to remove the file if it is used by any classloader and report that classloader to the user. We can try to do that in another PR, but for this reason I always pushed libraries inside ear/war files, because they have lifecycle while this is quite fragile.

Could it be enough to disable the application and then remove the library? Maybe we can move this to discussions.

avpinchuk commented:
> Maybe we can move this to discussions

Good idea ;)

OndroMih committed:
Update classloader in the add-library command - unit tests

And small changes based on review comments

OndroMih commented:
I used a wrong phase for the assembly plugin. It's supposed to copy test classes before classloader tests are executed. Before the fix, it was executed before test classes were compiled, it tried to move non-existent classes and failed.

OndroMih committed:
Update classloader in the add-library command - fix build

Skip creting assembly for tests if tests are no compiled in the fastest profile

OndroMih commented:
OK, the fastest profile disabled test compilation, so I needed to disable the assembly plugin in the fastest profile too.

dmatej commented on nucleus/core/kernel/src/test/java/com/sun/enterprise/v3/server/CommonClassLoaderServiceImplTest.java:
```diff
@@ -0,0 +1,191 @@
+        final String classesPath = "target/test-additional-classes/";
+        commonCLService.addToClassPath(new File(classesPath).toURI().toURL());
+
+// we need to retrieve the classloader after adding URLs, otherwise we
```
Is this aligned intentionally this way?

dmatej commented on nucleus/core/kernel/pom.xml:
```diff
@@ -260,6 +264,66 @@
             </plugin>
+            <plugin>
+                <artifactId>maven-assembly-plugin</artifactId>
+                <version>3.7.1</version>
```
Just remove the version, it is already set in parents.

dmatej reviewed (approved):
If we want to hurry to merge it to 7.0.18, those two comments can be resolved in another PR (or I can add it to mine as I am messing with classloaders too)
